test/e2e/e2e.go

```diff
@@ -1357,187 +1388,216 @@ var _ = ginkgo.Describe("e2e multiple ecmp external gateway validation", func()
+		annotateArgs := []string{
+			"annotate",
+			"namespace",
+			f.Namespace.Name,
+			fmt.Sprintf("k8s.ovn.org/routing-external-gws=%s,%s", addresses.gatewayIPs[0], addresses.gatewayIPs[1]),
 		}
+		framework.Logf("Annotating the external gateway test namespace to container gateways: %s, %s", addresses.gatewayIPs[0], addresses.gatewayIPs[1])
+		framework.RunKubectlOrDie(f.Namespace.Name, annotateArgs...)
 
 		// Verify the gateways and remote loopback addresses are reachable from the pod.
 		// Iterate checking connectivity to the loopbacks on the gateways until tcpdump see
 		// the traffic or 20 attempts fail. Odds of a false negative here is ~ (1/2)^20
-		ginkgo.By(fmt.Sprintf("Verifying ecmp connectivity to the external gateways by iterating through the prefix %s", exGWRemoteIpPrefix))
-
-		// create a buffered channel that will handle error reporting from the goroutines
-		icmpChan := make(chan error, 2)
+		ginkgo.By("Verifying ecmp connectivity to the external gateways by iterating through the targets")
 
 		// Check for egress traffic to both gateway loopback addresses using tcpdump, since
 		// /proc/net/dev counters only record the ingress interface traffic is received on.
 		// The test will waits until an ICMP packet is matched on the gateways or fail the
 		// test if a packet to the loopback is not received within the timer interval.
 		// If an ICMP packet is never detected, return the error via the specified chanel.
-		go func() {
-			_, err = runCommand("docker", "exec", gwContainer1, "timeout", testTimeout, "tcpdump", "-c", "1", "icmp")
-			if err == nil {
-				framework.Logf("ICMP packet successfully detected on gateway %s", gwContainer1)
-				icmpChan <- err
-			}
-			icmpChan <- err
-		}()
-		go func() {
-			_, err = runCommand("docker", "exec", gwContainer2, "timeout", testTimeout, "tcpdump", "-c", "1", "icmp")
-			if err == nil {
-				framework.Logf("ICMP packet successfully detected on gateway %s", gwContainer2)
-				icmpChan <- err
+
+		g := errgroup.Group{}
+		checkPingOnContainer := func(container string) error {
+			_, err := runCommand("docker", "exec", container, "timeout", testTimeout, "tcpdump", "-c", "1", icmpToDump)
+			if err != nil {
+				return err
 			}
-			icmpChan <- err
-		}()
+			framework.Logf("ICMP packet successfully detected on gateway %s", container)
+			return nil
+		}
+
+		g.Go(func() error { return checkPingOnContainer(gwContainer1) })
+		g.Go(func() error { return checkPingOnContainer(gwContainer2) })
 
 		// spawn a goroutine to asynchronously (to speed up the test)
 		// to ping the gateway loopbacks on both containers via ECMP.
-		for lastOctet := 1; lastOctet <= ecmpRetry; lastOctet++ {
-			gwLo := fmt.Sprintf("%s%d", exGWRemoteIpPrefix, lastOctet)
-			go func() {
-				_, err = framework.RunKubectl(f.Namespace.Name, "exec", srcPingPodName, testContainerFlag, "--", "ping", "-c", testTimeout, gwLo)
+		for _, address := range addresses.targetIPs {
+			go func(target string) {
```

Comment:
go routines inside ginkgo panic, check if we need to use `defer ginkgo.Recover()` here, I´m not sure if it applies here :smile: 
http://tonyyang132.blogspot.com/2018/08/ginkgo-and-goroutine.html

Reply:
I _think_ that happens after a failure. I'll add the defer, should not hurt.